plugins/synaprom/fu-synaprom-device.c

```diff
@@ -0,0 +1,504 @@
+				     "MFW metadata is invalid");
+		return NULL;
+	}
+	memcpy (&hdr, g_bytes_get_data (blob, NULL), sizeof(hdr));
+	if (hdr.product != FU_SYNAPROM_PRODUCT_PROMETHEUS) {
+		if (flags & FWUPD_INSTALL_FLAG_FORCE) {
+			g_warning ("MFW metadata not compatible, "
+				   "got 0x%02x expected 0x%02x",
+				   hdr.product, (guint) FU_SYNAPROM_PRODUCT_PROMETHEUS);
+		} else {
+			g_set_error (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_NOT_SUPPORTED,
+				     "MFW metadata not compatible, "
+				     "got 0x%02x expected 0x%02x",
+				     hdr.product, (guint) FU_SYNAPROM_PRODUCT_PROMETHEUS);
+			return NULL;
+		}
+	}
+	if (hdr.vmajor != self->vmajor || hdr.vminor != self->vminor) {
+		if (flags & FWUPD_INSTALL_FLAG_FORCE) {
+			g_warning ("MFW version not compatible, "
+				   "got %u.%u expected %u.%u",
+				   hdr.vmajor, hdr.vminor,
+				   self->vmajor, self->vminor);
+		} else {
+			g_set_error (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_NOT_SUPPORTED,
+				     "MFW version not compatible, "
+				     "got %u.%u expected %u.%u",
+				     hdr.vmajor, hdr.vminor,
+				     self->vmajor, self->vminor);
+			return NULL;
+		}
+	}
+
+	/* get payload */
+	return fu_synaprom_firmware_get_bytes_by_tag (firmware,
+						      FU_SYNAPROM_FIRMWARE_TAG_MFW_PAYLOAD,
+						      error);
+}
+
+static gboolean
+fu_synaprom_device_write_firmware (FuDevice *device, GBytes *fw, GError **error)
+{
+	FuSynapromDevice *self = FU_SYNAPROM_DEVICE (device);
+	const guint8 *buf;
+	gsize sz = 0;
+
+	/* write chunks */
+	fu_device_set_progress (device, 10);
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_WRITE);
+	buf = g_bytes_get_data (fw, &sz);
+	while (sz != 0) {
+		guint32 chunksz;
+		g_autoptr(GByteArray) chunk = g_byte_array_new ();
+
+		/* get chunk size */
+		if (sz < sizeof(guint32)) {
+			g_set_error_literal (error,
+					     G_IO_ERROR,
+					     G_IO_ERROR_NOT_SUPPORTED,
+					     "No enough data for patch len");
+			return FALSE;
+		}
+		memcpy (&chunksz, buf, sizeof(guint32));
+		buf += sizeof(guint32);
+		sz -= sizeof(guint32);
+		if (sz < chunksz) {
+			g_set_error_literal (error,
+					     G_IO_ERROR,
+					     G_IO_ERROR_NOT_SUPPORTED,
+					     "No enough data for patch chunk");
+			return FALSE;
+		}
+
+		/* download chunk */
+		g_byte_array_append (chunk, buf, chunksz);
+		if (!fu_synaprom_device_cmd_download_chunk (self, chunk, error))
+			return FALSE;
+
+		/* next chunk */
+		buf += chunksz;
+		sz -= chunksz;
+	}
+
+	/* success! */
+	fu_device_set_progress (device, 100);
+	return TRUE;
+}
+
+static gboolean
+fu_synaprom_device_probe (FuUsbDevice *device, GError **error)
+{
+	fu_device_add_flag (FU_DEVICE (device), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_set_remove_delay (FU_DEVICE (device), FU_DEVICE_REMOVE_DELAY_RE_ENUMERATE);
+	fu_device_set_name (FU_DEVICE (device), "Prometheus");
+	fu_device_set_summary (FU_DEVICE (device), "Fingerprint reader");
+	fu_device_set_vendor (FU_DEVICE (device), "Synaptics");
```

Comment:
Consider putting these into the quirk file instead?  Especially if it does turn out this plugin might be able to repurpose a lot of code for other non-prometheus model fingerprint readers from Synaptics.

Reply:
It's going to be prometheus-specific; I'm hopeful there'll be a long list of USB PIDs that all speak the same protocol that can be added to the quirk file.